[Just In Time Messages] Open WebView for JITM Call to Action

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #7855
<!-- Id number of the GitHub issue this PR addresses. -->
Not addressed: UTM parameters, analytic events

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
JITMs contain a link for the CTA button, which we should open in a webview to allow the merchant to take action on a given JITM.

This PR extracts the existing logic for opening a webview from the upsell card readers campaign, and generalises it into a `WooNavigationSheet`, used by `WebViewSheet`, which will contain either an `AuthenticatedWebView` or a `WebView`.

When the user taps the CTA for a message, the WebView is presented in a sheet, which can be dismissed with a swipe or by tapping Done. The title of the message is used for the WebViewSheet title to provide context.

woocommerce.com and wordpress.com links will be opened in an `AuthenticatedWebView`, which sets the appropriate authentication so that log in to these sites is more convenient. For users with 2FA, entering the code happens at the start of the flow, which isn't ideal but is a web behaviour, that the apps can't change (other than by using a non-authenticated webview.)

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### JITM on my store

N.B. see pdfdoF-1uc-p2#comment-2581 for details of setting up your store to be eligible for the test JITM

With a US store that is eligible for the test JITM, on a debug/alpha build of the app

1. Open the app
2. Observe that the message is shown
3. Tap the CTA
4. Observe that the webview is opened and takes you to the M2 reader page, and that when you log in you only need to authorise your wpcom account, not enter the credentials.

### Upsell Card Readers cards

The Upsell Card Reader campaigns have all been updated to use the extracted web views, so could do with being tested again.

In a US or CA WCPay store

1. Open the app and go to the Upsell Card Reader card
2. Tap the `Purchase card reader` button
3. Observe that the webview is opened and takes you to the M2/WisePad 3 reader page, and that when you log in you only need to authorise your wpcom account, not enter the credentials.

Locations of the card: top of Order List, and the Payment Method screen (after tapping Take Payment)

See #7862 for more detailed testing instructions.


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/2472348/198679425-5f28faba-b24b-4c13-84af-12a999916875.mp4



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

joshheald committed:
7855 Extract WebView creation from existing cards

joshheald committed:
7855 Show webview when JITM CTA is tapped

pachlava commented:
@joshheald 👋 

I can see that this is taking place on your video recording as well, so it's probably expected. I'm just wondering if authorizing the WPCOM account on _every_ opening of a webview has technical reasons which are hard to overcome? (if opening a webview is similar to starting a browser session from scratch every time, then it's understood)

pachlava commented:
I noticed no issues on iPhone, but for iPad there's one. The webview, when initiated from Stats (JITM) or Orders List (Upsell CR), has a split view:

<img width="454" alt="Screenshot 2022-10-31 at 10 39 52" src="https://user-images.githubusercontent.com/73365754/198966800-e8c44c12-cd6c-4d40-9218-000830a243da.png">

<img width="453" alt="Screenshot 2022-10-31 at 10 40 04" src="https://user-images.githubusercontent.com/73365754/198966824-6f8f0718-a840-4351-908a-b71a3bea79c0.png">

The viewbview looks perfect, though, when it's shown via collect payment flow (`IPP` / `Order Detail`):

<img width="451" alt="Screenshot 2022-10-31 at 10 43 03" src="https://user-images.githubusercontent.com/73365754/198967278-7d3eb424-9f72-443a-bc3d-7d4b90a61ee8.png">

joshheald commented:
> I noticed no issues on iPhone, but for iPad there's one. The webview, when initiated from Stats (JITM) or Orders List (Upsell CR), has a split view:

@pachlava Thanks, good catch, and that's a strange one! I suspect it's because it's being presented from a Split View in the Orders and Dashboard cases, but on payment methods it's presented from another modal. I'll dig in to it.

> I'm just wondering if authorizing the WPCOM account on every opening of a webview has technical reasons which are hard to overcome? (if opening a webview is similar to starting a browser session from scratch every time, then it's understood)

I've not really tried to overcome the issues here, but my memory and understanding of web views is that we can't rely on keeping the session around between launches. It's potentially worth another project to dig in to the detail here and see whether we can avoid doing it every time, but I think that's outside the scope of this work.

joshheald committed:
StackNavigationViewStyle avoids split view on iPad

Previously when the sheet was created on iPad, the content would be rendered as the master view of a split view, with no detail view. Often this would lead to no content being displayed, and a confusing back button added.

StackNavigationViewStyle makes no change to the iPhone design but makes this sheet suitable for form/page sheet presentation on iPad, showing the content immediately.

joshheald commented:
@pachlava iPad display issue fixed in c319e23. 

Thanks for finding this – it is a pre-existing issue in the Order List.

https://user-images.githubusercontent.com/2472348/198993856-00797362-71f3-4dc8-8fa3-cba1d9bb721c.mp4

pachlava commented:
> @pachlava iPad display issue fixed in https://github.com/woocommerce/woocommerce-ios/commit/c319e23253915e64fd9509a241fa7787284e3703.

Thanks, @joshheald - I'm taking a look

pachlava reviewed (approved):
No complaints from my side, @joshheald: the fix for webview layout on iPda looks good to me 👍 🙂 Thanks!

One question: is it by design that webview title initiated from JITM is different from the one initiated from Upsell CR?

JITM:
<img width="325" alt="Screenshot 2022-10-31 at 13 32 26" src="https://user-images.githubusercontent.com/73365754/198999070-5481aa96-09b6-4304-92ea-f3c930d3f4f6.png">

UCR:
<img width="325" alt="Screenshot 2022-10-31 at 13 37 40" src="https://user-images.githubusercontent.com/73365754/198999267-6f965baa-9c3e-4ab0-80c0-d7532fcad1e3.png">

joshheald commented:
> One question: is it by design that webview title initiated from JITM is different from the one initiated from Upsell CR?

@pachlava thanks for checking again so quickly.

hat is using the title of the JITM (i.e. the bold text displayed at the top of the JITM card) for the title of the webview too. Since the message could be anything, and we only have a limited number of fields to use, when I wrote it that seemed like the best one. The UCR campaign doesn't have this limitation....

However, I could use the button title instead, which actually would match the UCR implementation. I'll check what Anirudh did on Android, if anything, and mirror it in a future PR

pachlava commented:
Both UCR and JITM webivews on Android use `Order Card Reader` for title. I think it would be nice to have it aligned on iOS, because both are eventually about the same thing.

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/WebView.swift:
```diff
@@ -0,0 +1,65 @@
+import class Networking.UserAgent
+
+/// Mirror of AuthenticatedWebView, for equivalent display of URLs in `WKWebView` that do not need authentication on WPCom.
+struct WebView: UIViewRepresentable {
```
I am wondering, since most of this code is duplicated with `AuthenticatedWebView`, is there a way to collapse both structs into one with some sort of configuration to authenticate when necessary?

joshheald replied:
@toupper Probably... I'll take a look towards the end of the week in another PR, as I'm trying to de-risk delivery in 11.1 as much as possible. Other than that, would you be happy to approve?

toupper replied:
That's great! sure thing 👍

joshheald committed:
Merge branch 'trunk' into issue/7855-open-webview-for-jitm-cta

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewController.swift
#	WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift
